File: communication/Switchboard.py
"""
Switchboard Class

The main switchboard for the project, managing callbacks when events occur.
"""

from Events import Events
from Queue import Queue

from threading import Lock

# Protects the switchboard object
mutex = Lock()

class Switchboard :
    def __init__(self, dustbin) :
        self.switchboard = { k : {} for k in Events.KEYS}
        self.DUSTBIN = dustbin
        self.toTrigger = Queue()
        self.keepGoing = True
    def stop(self) :
        self.DUSTBIN.log('ENDING SWITCHBOARD THREAD')
        self.keepGoing = False
        self.toTrigger.put((None,None))
    def unsubscribe(self, event, listener) :
        mutex.acquire()
        if listener.getHash() in self.switchboard[event] :
            del self.switchboard[event][listener.getHash()]
        mutex.release()
    def subscribe(self, event, listener, callback=None) :
        mutex.acquire()
        self.DUSTBIN.log('Setting callback to event %d' %event)
        self.switchboard[event][listener.getHash()] = listener
        if callback is not None :
            callback()
        mutex.release()
    def _triggerEvent(self, event, kwargs) :
        if event is None and kwargs is None :
            return
        mutex.acquire()
        self.DUSTBIN.log("Event %d triggered." %event)
        for listener in self.switchboard[event].values() :
            try :
                listener.runCallback(kwargs)
            except Exception as e :
                # Program should not die.
                self.DUSTBIN.log('EXCEPTION IN CALLBACK FOR EVENT', e, e.message)
                continue
        mutex.release()
    def run(self) :
        try :
            while self.keepGoing :
                self.DUSTBIN.log('Switchboard process continues.')
                event, kwargs = self.toTrigger.get(True)
                self._triggerEvent(event, kwargs)
        except Exception as e :
            self.DUSTBIN.log('Switchboard process had a fatal error.', e)
            print('Switchboard process had a fatal error.', e)
        self.DUSTBIN.log('SWITCHBOARD OFFICIALLY DEAD.')
    def runTrigger(self, event, kwargs) :
        self.toTrigger.put((event, kwargs))

File: communication/Switchboard.js
'use strict';

/*
  Switchboard Class
  The main switchboard for the project, managing callbacks when events occur.
*/

const {Events, EventListener} = require('./Events.js');
const Queue = require('../Queue.js');

class Switchboard {
  constructor(_DUSTBIN) {
    this.switchboard = {};
    for (let k of Object.values(Events)) {
      this.switchboard[k] = {};
    }
    this.toTrigger = new Queue(); 
    this.DUSTBIN = _DUSTBIN;
    this.keepGoing = true;
  }
  stop() {
    this.DUSTBIN.log('ENDING SWITCHBOARD THREAD');
    this.keepGoing = false;
    // Let's try to fix this so that we don't need some dumb extra null in the queue.
    this.toTrigger.put({event: null, kwargs: null});
  } 
  // Returns unsubscribe function.
  subscribe(event, listener, callback=null) {
    this.DUSTBIN.log(`Setting callback to event ${event}`);
    this.switchboard[event][listener.getHash()] = listener;
    if (callback) {
      callback();
    }
    return ()=>{this.swithcboard = this.switchboard.filter(l => l != listener)}
  }
  
  runTrigger(_event, _kwargs) {
    this.toTrigger.put({event: _event, kwargs: _kwargs});
  }
  _triggerEvent(event, kwargs) {
    if (!event && !kwargs) {
      return;
    }
    this.DUSTBIN.log(`Event ${event} triggered.`)
    for (let listener in Object.values(this.switchboard[event])) {
      try {
        listener.runCallback(kwargs);
      }
      catch (e) {
        // Program should not die.
        this.DUSTBIN.log('EXCEPTION IN CALLBACK FOR EVENT', e, e.message);
        continue;
      }
    }
  }
  run() {
    try {
      while (this.keepGoing) {
        this.DUSTBIN.log('Switchboard process continues.');
        let {event, kwargs} = this.toTrigger.pop();
        this._triggerEvent(event, kwargs);
      }
    } catch (e) {
      this.DUSTBIN.log('Switchboard process had a fatal error.', e);
      console.log('Switchboard process had a fatal error.', e);
    }
    this.DUSTBIN.log('SWITCHBOARD OFFICIALLY DEAD.');
  }
}

module.exports = Switchboard;